File: src/game/morgue.ts
import type { PuzzleData } from '../types';

/**
 * The paper's morgue: its own clippings, searched by a word (docs/SOUL-FEATURES.md,
 * idea 27, after *Her Story*). Only pages the agent has decoded answer, so the
 * search can never reveal a line, or even confirm a guess, from a page still open.
 * Android's `Morgue` answers the same way.
 */
export const MORGUE_MIN_QUERY = 3;

export function searchMorgue(puzzles: PuzzleData[], solvedPuzzleIds: string[], query: string): PuzzleData[] {
  const needle = query.trim().toUpperCase();
  if (needle.length < MORGUE_MIN_QUERY) return [];
  const solved = new Set(solvedPuzzleIds);
  return puzzles.filter(
    (puzzle) =>
      solved.has(puzzle.id) &&
      `${puzzle.originalText} ${puzzle.headline} ${puzzle.subheadline}`.toUpperCase().includes(needle)
  );
}

File: android/core/cipher/src/main/kotlin/com/chroniclecryptogram/cipher/Morgue.kt
package com.chroniclecryptogram.cipher

import com.chroniclecryptogram.cipher.model.PuzzleData

/**
 * The paper's morgue: its own clippings, searched by a word (docs/SOUL-FEATURES.md,
 * idea 27). Only pages the agent has decoded answer, so the search can never reveal a
 * line, or confirm a guess, from a page still open. The web's `game/morgue.ts`
 * answers the same way.
 */
object Morgue {

    const val MIN_QUERY = 3

    fun search(puzzles: List<PuzzleData>, solvedPuzzleIds: Set<String>, query: String): List<PuzzleData> {
        val needle = query.trim().uppercase()
        if (needle.length < MIN_QUERY) return emptyList()
        return puzzles.filter { puzzle ->
            puzzle.id in solvedPuzzleIds &&
                "${puzzle.originalText} ${puzzle.headline} ${puzzle.subheadline}".uppercase().contains(needle)
        }
    }
}
